### ISSUE TITLE
Allow auto-pairing Bluetooth joypads

### ISSUE THREAD

**Issue by hadess** [libretro/RetroArch#2962]
For an "embedded" version of Retroarch, it would be useful if it was possible for RetroArch to look for joypads in the vicinity to pair with them, when it doesn't have any plugged in/connected.

When no joypads are connected, Bluetooth is available, and the feature is turned on, you would:
- turn on discoverable for the Bluetooth adapter
- start looking for joypads in pairing modes
- pair with them
- show that a new device was paired (OSD?)

You would stop the process as soon as a joypad was connected, via USB or an already configured device connected via Bluetooth.

Note that this could also implementable through a (configurable) helper. If that were the case, I would probably write my own using glib/D-Bus (I'm the maintainer of gnome-bluetooth).

**Comment by Kivutar** [comment]
I remember that Twinaphex had some plans for bluetooth in RA. Do you have any advice on how to implement such pairing in the most portable way? RA supports a lot of platforms and avoiding dependancies, or choosing portable dependancies, is important if we want things to work the same on Wii, iOS, Android, Linux, etc.

**Comment by hadess** [comment]
It would be much like the joypad code, with each system having to have specific code written for that. Contrary to the joypad code though, it's likely to be involved and complicated if you want to minimise external library dependencies. I'm not sure it's worth it at this point.

Which is why I was interested in the "helper" way of implementing this. I'd be happy writing a reference implementation for non-Android Linux, based on bluez 5. This would give me an excuse to buy a Bluetooth joypad ;)

**Comment by twinaphex** [comment]
Yes, this is definitely something that has to happen. If you want to start such an initiative, by all means.

There might be some now-dead iOS code we might have to refactor too then to fit this new driver model - 

https://github.com/libretro/RetroArch/blob/master/input/drivers_hid/btstack_hid.c

This was a BTStack implementation that allowed for Bluetooth gamepad support on iOS. I think it has long since stopped working though since you know how Apple is with their OS upgrades and their forced deprecation. However, there might still be some useful parts here that we could use for constructing this new driver interface for wireless input.

**Comment by hooby3dfx** [comment]
here is a script which does all of the OPs suggestions which can be run on Lakka on boot:
https://github.com/hooby3dfx/lakka-addons/blob/master/btautopair.sh
i know its not the real solution for retroarch, but its a pretty good workaround for now!

**Comment by Luraktinus** [comment]
Having to log in via ssh just to connect a gamepad is painful as hell. Add this or at least a Bluetooth menu like retropie has

**Comment by gouchi** [comment]
Link to bounty  #8916

**Comment by gouchi** [comment]
@hadess This issue should  be closed as we have now some [bluetooth drivers](https://github.com/libretro/RetroArch/tree/master/bluetooth/drivers) ;-)
